encoder.go

```diff
@@ -38,67 +46,93 @@ func newMaxEncoder() *encoder {
+func encodeNative[T native](val T, t wafObjectType, obj *wafObject) {
 	obj._type = t
 	obj.value = (uintptr)(val)
-	return nil
 }
 
-func (encoder *encoder) encode(value reflect.Value, obj *wafObject, depth int) error {
+var nullableTypeKinds = map[reflect.Kind]struct{}{
+	reflect.Interface:     {},
+	reflect.Pointer:       {},
+	reflect.UnsafePointer: {},
+	reflect.Map:           {},
+	reflect.Slice:         {},
+	reflect.Func:          {},
+	reflect.Chan:          {},
+}
+
+// isValueNil check if the value is nullable and if it is actually nil
+// we cannot directly use value.IsNil() because it panics on non-pointer values
+func isValueNil(value reflect.Value) bool {
+	_, nullable := nullableTypeKinds[value.Kind()]
+	return nullable && value.IsNil()
+}
+
+func (encoder *encoder) encode(value reflect.Value, obj *wafObject, depth int) {
+	value.IsValid()
 	switch kind := value.Kind(); {
 	// Terminal cases (leafs of the tree)
 	case kind == reflect.Invalid:
-		return errUnsupportedValue
+		encodeNative[uintptr](0, wafInvalidType, obj)
+	// Is nullable type kind
+	case isValueNil(value):
+		encodeNative[uintptr](0, wafNilType, obj)
+
 	// 		Booleans
 	case kind == reflect.Bool:
-		return encodeNative(nativeToUintptr(value.Bool()), wafBoolType, obj)
+		encodeNative(nativeToUintptr(value.Bool()), wafBoolType, obj)
 	// 		Numbers
 	case value.CanInt(): // any int type or alias
-		return encodeNative(value.Int(), wafIntType, obj)
+		encodeNative(value.Int(), wafIntType, obj)
 	case value.CanUint(): // any Uint type or alias
-		return encodeNative(value.Uint(), wafUintType, obj)
+		encodeNative(value.Uint(), wafUintType, obj)
 	case value.CanFloat(): // any float type or alias
-		return encodeNative(nativeToUintptr(value.Float()), wafFloatType, obj)
+		encodeNative(nativeToUintptr(value.Float()), wafFloatType, obj)
 
 	//		Strings
 	case kind == reflect.String: // string type
-		return encoder.encodeString(value.String(), obj)
+		encoder.encodeString(value.String(), obj)
 	case value.Type() == reflect.TypeOf([]byte(nil)): // byte array -> string
-		return encoder.encodeString(string(value.Bytes()), obj)
+		encoder.encodeString(string(value.Bytes()), obj)
 
 	// Recursive cases (internal nodes of the tree)
 	case kind == reflect.Interface || kind == reflect.Pointer: // Pointer and interfaces are not taken into account
-		return encoder.encode(value.Elem(), obj, depth)
+		encoder.encode(value.Elem(), obj, depth)
 	case kind == reflect.Array || kind == reflect.Slice: // either an array or a slice of an array
-		return encoder.encodeArray(value, obj, depth)
+		encoder.encodeArray(value, obj, depth)
 	case kind == reflect.Map:
-		return encoder.encodeMap(value, obj, depth)
+		encoder.encodeMap(value, obj, depth)
 	case kind == reflect.Struct:
-		return encoder.encodeStruct(value, obj, depth)
+		encoder.encodeStruct(value, obj, depth)
 
 	default:
-		return errUnsupportedValue
+		encodeNative[uintptr](0, wafInvalidType, obj)
```

Comment:
The case where `kind == reflect.Invalid` has the exact same code as the `default` case... They could be de-duplicated.

Reply:
Not mentioning that the `default` is also future-proof, the real reason we still keep the `kind == reflect.Invalid` case is because every other call on the value will potentially crash as the doc of reflect states that "each function that can be called with invalid values are explicitly stated as such, other functions do not support these values"